### DISCUSSION TITLE
New Types in UP

### DISCUSSION THREAD

**Discussion by JoanEspasa** [aiplan4eu/unified-planning#508] [Q&A] [answered]
Hi!

I'm interested in adding specific types to UP and then take care of them in the solver side. The usual case of PDDL-with-a-small-language-extension. For example, adding rationals, complex numbers or sets. What is the canonical way of doing such a thing, if possible?

Thanks,

Joan

**Comment by Framba-Luca** [comment]
Hi @JoanEspasa ,

Rationals are supported at UP level (we use `Fraction`), but in PDDL they are dumped as float.

Adding something else, like complex numbers, isn't really covered and can't be done in an "easy" way; if not by explicitly adding the support.

We will consider what needs to be done in order to support a general use-case like yours.

**Comment by JoanEspasa** [comment]
Thanks @Framba-Luca , it's good to know that it's not supported out-of-the-box as I suspected.

To give you a further example, consider a logistics-style domain where trucks are required to deliver raw materials between locations. Materials are loaded into trucks one big "pack" at a time, but all materials in a truck are unloaded at the same time, as they are dump-trucks. These problems get very naturally modelled as sets. 

My ideal scenario would be to add sets for example as a type for any Python API user, allow them to use them to construct problems and then translate them away via a Compiler. As I have no knowledge of UP internals, I'm wondering what is the amount of effort you expect such feature would entail, or if it is even viable with the current design.

**Reply by Framba-Luca** [reply]
To me this use-case resembles  discussion #44 ; could what you are saying here, be a specific case of the resource representation?

**Reply by Framba-Luca** [reply]
As for the expected effort, I expect this to be quite a complex task, considering the compiler and the correct behavior with respect to all that's already there.

**Reply by JoanEspasa** [reply]
#44 is similar to what I'm trying to achieve, yes: adding a new "thing" that can appear both in preconditions and effects. More concretely, I reckon that it would entail much more than just adding a new class in https://github.com/aiplan4eu/unified-planning/blob/master/unified_planning/model/types.py that inherits from the Type class, right?

**Reply by Framba-Luca** [reply, marked as answer]
At glance I would say you have to create a new _FNode_ that contains the new type, then specify how the new _FNode_ interacts with other types on every operand (in the _unified_planning/model/walkers/type_checker.py_), add checks around the code that the type is not the new one (where it can't be) and, in every walker, specify how it should behave on the new expression.

So, yes, adding complex numbers (for example) is not trivial and requires knowledge about how every walker should behave. Some might be trivial and might not even need a new function (the default one works), but others require effort and testing.

Adding sets would be like #44, which I am not certain what would be needed in order to fully support them.

**Reply by JoanEspasa** [reply]
Thanks for the pointers, I'll have a proper look at the source and give it a go :)
